File: Examples/cxx/Tests/source/ObservationsTests.cpp
#include "ObservationsTests.h"

using namespace Elektronik::Common::Data::Pb;

void ObservationsTests::SetUp()
{
    TestsBase::SetUp();

    std::vector<PointPb> pts;

    for (size_t i = 0; i < 5; ++i) {
        PointPb pt;
        pt.set_id(i);
        pt.set_message(std::to_string(i));
        pts.push_back(pt);
    }

    Vector3Pb pos; pos.set_x(0); pos.set_y(.5f);
    pts[0].set_allocated_position(new Vector3Pb(pos));

    pos.Clear(); pos.set_x(.5f); pos.set_y(-.5f);
    pts[1].set_allocated_position(new Vector3Pb(pos));

    pos.Clear(); pos.set_x(-.5f); pos.set_y(-.5f);
    pts[2].set_allocated_position(new Vector3Pb(pos));

    pos.Clear(); pos.set_x(-.5f);
    pts[3].set_allocated_position(new Vector3Pb(pos));

    pos.Clear(); pos.set_x(.5f);
    pts[4].set_allocated_position(new Vector3Pb(pos));

    for (const auto& pt : pts) {
        ObservationPb obs;
        obs.set_allocated_point(new PointPb(pt));
        m_map.push_back(obs);
    }

    ConnectionPb connection;

    connection.set_id1(m_map[0].point().id()); connection.set_id2(m_map[1].point().id());
    m_connections.push_back(connection);

    connection.set_id1(m_map[0].point().id()); connection.set_id2(m_map[2].point().id());
    m_connections.push_back(connection);

    connection.set_id1(m_map[2].point().id()); connection.set_id2(m_map[4].point().id());
    m_connections.push_back(connection);

    connection.set_id1(m_map[1].point().id()); connection.set_id2(m_map[3].point().id());
    m_connections.push_back(connection);

    connection.set_id1(m_map[3].point().id()); connection.set_id2(m_map[4].point().id());
    m_connections.push_back(connection);
}

TEST_F(ObservationsTests, Create) {
    auto observations = new PacketPb_Observations();
    for (const auto pt : m_map) {
        auto newPt = observations->add_data();
        newPt->CopyFrom(pt);
    }
    PacketPb packet;
    packet.set_action(PacketPb_ActionType::PacketPb_ActionType_add);
    packet.set_allocated_observations(observations);

    grpc::ClientContext ctx;
    ErrorStatusPb status;
    m_client->Handle(&ctx, packet, &status);

    ASSERT_EQ(status.err_type(), ErrorStatusPb_ErrorStatusEnum::ErrorStatusPb_ErrorStatusEnum_SUCCEEDED);
}

TEST_F(ObservationsTests, Update) {
    auto observations = new PacketPb_Observations();
    for (const auto obs : { m_map[2], m_map[4] }) {
        auto position = new Vector3Pb(obs.point().position());
        position->set_z(position->z() + .5f);

        auto point = new PointPb(obs.point());
        point->set_allocated_position(position);

        auto newObs = observations->add_data();
        newObs->CopyFrom(obs);
        newObs->set_allocated_point(point);
    }
    PacketPb packet;
    packet.set_action(PacketPb_ActionType::PacketPb_ActionType_update);
    packet.set_allocated_observations(observations);

    grpc::ClientContext ctx;
    ErrorStatusPb status;
    m_client->Handle(&ctx, packet, &status);

    ASSERT_EQ(status.err_type(), ErrorStatusPb_ErrorStatusEnum::ErrorStatusPb_ErrorStatusEnum_SUCCEEDED);
}

TEST_F(ObservationsTests, UpdateConnections) {
    PacketPb packet;
    packet.set_action(PacketPb_ActionType::PacketPb_ActionType_update);
    packet.set_allocated_observations(new PacketPb_Observations());

    auto connections = new PacketPb_Connections();
    connections->set_action(PacketPb_Connections_Action::PacketPb_Connections_Action_add);
    for (const auto c : m_connections) {
        auto c_ = connections->add_data();
        c_->CopyFrom(c);
    }
    packet.set_allocated_connections(connections);

    grpc::ClientContext ctx;
    ErrorStatusPb status;
    m_client->Handle(&ctx, packet, &status);

    ASSERT_EQ(status.err_type(), ErrorStatusPb_ErrorStatusEnum::ErrorStatusPb_ErrorStatusEnum_SUCCEEDED);
}

TEST_F(ObservationsTests, RemoveConnections) {
    PacketPb packet;
    packet.set_action(PacketPb_ActionType::PacketPb_ActionType_update);
    packet.set_allocated_observations(new PacketPb_Observations());

    auto connections = new PacketPb_Connections();
    connections->set_action(PacketPb_Connections_Action::PacketPb_Connections_Action_remove);
    for (const auto& c : { m_connections[0], m_connections[1], }) {
        auto connection = connections->add_data();
        connection->CopyFrom(c);
    }
    packet.set_allocated_connections(connections);

    grpc::ClientContext ctx;
    ErrorStatusPb status;
    m_client->Handle(&ctx, packet, &status);

    ASSERT_EQ(status.err_type(), ErrorStatusPb_ErrorStatusEnum::ErrorStatusPb_ErrorStatusEnum_SUCCEEDED);
}

TEST_F(ObservationsTests, Remove) {
    PacketPb packet;
    packet.set_action(PacketPb_ActionType::PacketPb_ActionType_remove);
    auto pts2remove = new PacketPb_Observations();
    for (const auto& obs : { m_map[1], m_map[3] }) {
        auto pt_ = pts2remove->add_data();
        pt_->CopyFrom(obs);
    }
    packet.set_allocated_observations(pts2remove);

    grpc::ClientContext ctx;
    ErrorStatusPb status;
    m_client->Handle(&ctx, packet, &status);

    ASSERT_EQ(status.err_type(), ErrorStatusPb_ErrorStatusEnum::ErrorStatusPb_ErrorStatusEnum_SUCCEEDED);
}

TEST_F(ObservationsTests, Clear) {
    PacketPb packet;
    packet.set_action(PacketPb_ActionType::PacketPb_ActionType_clear);
    packet.set_allocated_observations(new PacketPb_Observations());

    grpc::ClientContext ctx;
    ErrorStatusPb status;
    m_client->Handle(&ctx, packet, &status);

    ASSERT_EQ(status.err_type(), ErrorStatusPb_ErrorStatusEnum::ErrorStatusPb_ErrorStatusEnum_SUCCEEDED);
}

File: Examples/csharp/csharp.Tests/ObservationsTests.cs
using Elektronik.Common.Data.Pb;
using NUnit.Framework;
using System;
using System.Linq;


namespace csharp.Tests
{
    public class ObservationsTests : TestsBase
    {
        private ObservationPb[] m_map;
        private ConnectionPb[] m_connections;

        public ObservationsTests()
        {
            m_map = Enumerable.Range(0, 5).Select(id => new ObservationPb()
            {
                Point = new PointPb() 
                {
                    Id = id,
                    Message = $"{id}",
                }
            }).ToArray();

            m_map[0].Point.Position = new Vector3Pb() { X = 0, Y = .5 };
            m_map[1].Point.Position = new Vector3Pb() { X = .5, Y = -.5 };
            m_map[2].Point.Position = new Vector3Pb() { X = -.5, Y = -.5 };
            m_map[3].Point.Position = new Vector3Pb() { X = -.5, Y = 0 };
            m_map[4].Point.Position = new Vector3Pb() { X = .5, Y = 0 };

            m_connections = new[]
            {
                new ConnectionPb() { Id1 = m_map[0].Point.Id, Id2 = m_map[1].Point.Id, },
                new ConnectionPb() { Id1 = m_map[0].Point.Id, Id2 = m_map[2].Point.Id, },
                new ConnectionPb() { Id1 = m_map[2].Point.Id, Id2 = m_map[4].Point.Id, },
                new ConnectionPb() { Id1 = m_map[1].Point.Id, Id2 = m_map[3].Point.Id, },
                new ConnectionPb() { Id1 = m_map[3].Point.Id, Id2 = m_map[4].Point.Id, },
            };
        }

        [Test, Order(1)]
        public void Create()
        {
            var packet = new PacketPb()
            {
                Action = PacketPb.Types.ActionType.Add,
                Observations = new PacketPb.Types.Observations(),
            };
            var obs = Enumerable.Range(0, 5).Select(id => new ObservationPb()
            {
                Point = new PointPb()
                {
                    Id = id,
                    Message = $"{id}",
                }
            }).ToArray();
            obs[0].Point.Position = new Vector3Pb() { X = 0, Y = .5 };
            obs[1].Point.Position = new Vector3Pb() { X = .5, Y = -.5 };
            obs[2].Point.Position = new Vector3Pb() { X = -.5, Y = -.5 };
            obs[3].Point.Position = new Vector3Pb() { X = -.5, Y = 0 };
            obs[4].Point.Position = new Vector3Pb() { X = .5, Y = 0 };
            packet.Observations.Data.Add(obs);

            var response = m_client.Handle(packet);
            Assert.True(response.ErrType == ErrorStatusPb.Types.ErrorStatusEnum.Succeeded, response.Message);
        }

        [Test, Order(2)]
        public void Update()
        {
            var packet = new PacketPb()
            {
                Action = PacketPb.Types.ActionType.Update,
                Observations = new PacketPb.Types.Observations(),
            };
            m_map[0].Point.Position.Z += .5;
            m_map[2].Point.Position.Z += .5;
            m_map[4].Point.Position.Z += .5;

            packet.Observations.Data.Add(m_map);

            var response = m_client.Handle(packet);
            Assert.True(response.ErrType == ErrorStatusPb.Types.ErrorStatusEnum.Succeeded, response.Message);
        }

        [Test, Order(3)]
        public void UpdateConnections()
        {
            var packet = new PacketPb()
            {
                Action = PacketPb.Types.ActionType.Update,
                Observations = new PacketPb.Types.Observations(),
                Connections = new PacketPb.Types.Connections()
                {
                    Action = PacketPb.Types.Connections.Types.Action.Add,
                },
            };
            packet.Connections.Data.Add(m_connections);
            var response = m_client.Handle(packet);
            Assert.True(response.ErrType == ErrorStatusPb.Types.ErrorStatusEnum.Succeeded, response.Message);
        }

        [Test, Order(4)]
        public void RemoveConnections()
        {
            var packet = new PacketPb()
            {
                Action = PacketPb.Types.ActionType.Update,
                Observations = new PacketPb.Types.Observations(),
                Connections = new PacketPb.Types.Connections()
                {
                    Action = PacketPb.Types.Connections.Types.Action.Remove,
                },
            };
            packet.Connections.Data.Add(new[] { m_connections[0], m_connections[1] });
            var response = m_client.Handle(packet);
            Assert.True(response.ErrType == ErrorStatusPb.Types.ErrorStatusEnum.Succeeded, response.Message);
        }

        [Test, Order(5)]
        public void Remove()
        {
            var packet = new PacketPb()
            {
                Action = PacketPb.Types.ActionType.Remove,
                Observations = new PacketPb.Types.Observations(),
            };

            packet.Observations.Data.Add(new[] { m_map[1], m_map[3] });

            var response = m_client.Handle(packet);
            Assert.True(response.ErrType == ErrorStatusPb.Types.ErrorStatusEnum.Succeeded, response.Message);
        }

        [Test, Order(6)]
        public void Clear()
        {
            var packet = new PacketPb()
            {
                Action = PacketPb.Types.ActionType.Clear,
                Observations = new PacketPb.Types.Observations(),
            };

            var response = m_client.Handle(packet);
            Assert.True(response.ErrType == ErrorStatusPb.Types.ErrorStatusEnum.Succeeded, response.Message);
        }
    }
}